tests/components/shelly/conftest.py

```diff
@@ -104,16 +117,52 @@ async def coap_wrapper(hass):
         blocks=MOCK_BLOCKS,
         settings=MOCK_SETTINGS,
         shelly=MOCK_SHELLY,
+        firmware_version="some fw string",
+        update=AsyncMock(),
+        initialized=True,
+    )
+
+    hass.data[DOMAIN] = {DATA_CONFIG_ENTRY: {}}
+    hass.data[DOMAIN][DATA_CONFIG_ENTRY][config_entry.entry_id] = {}
+    wrapper = hass.data[DOMAIN][DATA_CONFIG_ENTRY][config_entry.entry_id][
+        BLOCK
+    ] = BlockDeviceWrapper(hass, config_entry, device)
+
+    wrapper.async_setup()
+
+    return wrapper
+
+
+@pytest.fixture
+async def rpc_wrapper(hass):
+    """Setups a coap wrapper with mocked device."""
+    await async_setup_component(hass, "shelly", {})
+
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={"sleep_period": 0, "model": "SNSW-001P16EU", "gen": 2},
+        unique_id="12345678",
+    )
+    config_entry.add_to_hass(hass)
+
+    device = Mock(
+        call_rpc=AsyncMock(),
+        config=MOCK_CONFIG,
+        event={},
+        shelly=MOCK_SHELLY,
+        status=MOCK_STATUS,
+        firmware_version="some fw string",
         update=AsyncMock(),
         initialized=True,
+        shutdown=AsyncMock(),
     )
 
     hass.data[DOMAIN] = {DATA_CONFIG_ENTRY: {}}
     hass.data[DOMAIN][DATA_CONFIG_ENTRY][config_entry.entry_id] = {}
     wrapper = hass.data[DOMAIN][DATA_CONFIG_ENTRY][config_entry.entry_id][
-        COAP
-    ] = ShellyDeviceWrapper(hass, config_entry, device)
+        RPC
+    ] = RpcDeviceWrapper(hass, config_entry, device)
 
-    await wrapper.async_setup()
+    wrapper.async_setup()
```

Comment:
It's not a good pattern to access integration details like hass.data or the coordinator in the tests. We should aim for just setting up the integration while patching appropriately. This will make the tests more robust.